[MTIA Aten Backend][2/n] Migrate clamp ops(clamp.out/clamp_min.out/clamp_max.out) from out-of-tree to in-tree

andyanwang opened the pull request:
Summary:
# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This PR 
1. Migrate 3 clamp ops from out-of-tree to in-tree(had to migrate the 3 ops altogether, because clamp.out calls all 3 stubs, which are also called by the other 2 ops):
- clamp.out
- clamp_min.out
- clamp_max.out
2. Also enabled structured kernel codegen for MTIA, which is needed by clamp
3. Also introduced the `--mtia` flag to torchgen to prevent OSS from gencoding MTIA code.(Otherwise we got such link error `lib/libtorch_cpu.so: undefined reference to at::detail::empty_mtia`)

Differential Revision: D74674418




cc @egienvalue @albanD

andyanwang commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -1535,7 +1535,6 @@ void TensorIteratorBase::build(TensorIteratorConfig& config) {
   // Nothing beyond this point is important for meta functions, so it's fine to exit early here.
   // Extend the condition to MAIA tesnors as MAIA tensors also don't have storage.
   if (privateuse1_without_storage  ||
-      common_device_.type() == DeviceType::MTIA ||
```
I am not sure if this line should be deleted. I'm deleting it to reach parity with CPU and CUDA. But our testing actually works no matter whether this line is deleted or not.

albanD replied:
Sounds good if CI is green!

andyanwang commented:
This PR is tested using mtia unit tests, as shown in the test plan of the internal diff D74674418

andyanwang commented:
@pytorchbot label "topic: not user facing"

andyanwang commented on aten/src/ATen/native/mtia/EmptyTensor.cpp:
```diff
@@ -0,0 +1,83 @@
+  return GetAllocator(DeviceType::MTIA);
+}
+
+TensorBase empty_mtia(
```
We are adding the empty_mtia and empty_strided_mtia apis as they are required by pytorch structured kernel generated code.

albanD replied:
You should be able to register these as the impl for empty/empty_strided in a follow up PR !

andyanwang replied:
I'll look into that!

andyanwang commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -84,6 +84,9 @@ file(GLOB mkldnn_xpu_h "native/mkldnn/xpu/*.h" "native/mkldnn/xpu/detail/*.h")
 
 file(GLOB native_cpp "native/*.cpp")
+# Exclude MTIA files from native_cpp
+file(GLOB native_mtia_cpp "native/mtia/*.cpp")
```
If not exclude mtia files, we got such link errors:
```
/usr/bin/ld: lib/libtorch_cpu.so: undefined reference to `at::detail::empty_mtia(c10::ArrayRef<long>, c10::TensorOptions const&)'
```

andyanwang commented:
Note there're CI failures: 
```
/usr/bin/ld: lib/libtorch_cpu.so: undefined reference to `at::detail::empty_mtia(c10::ArrayRef<long>, c10::TensorOptions const&)'
```

It seems it's because the generated files `RegisterMTIA_0.cpp` or `RegisterMTIAEverything.cpp` was included in CMake build.  Anybody knows how can can we exclude these files from libtorch_cpu build? cc @albanD

nautsimon reviewed (commented):
Need to add a couple things to `torchgen/gen.py` to resolve the build issues

```
# Line ≈2824
    parser.add_argument(
        "--mtia",
        action="store_true",
        help="Generate MTIA registration code when set",
    )
```


```
# line ≈2925
if not options.mtia:
        ignore_keys.add(DispatchKey.MTIA)

        if DispatchKey.MTIA in dispatch_keys:
            del dispatch_keys[dispatch_keys.index(DispatchKey.MTIA)]
```

andyanwang commented on build.bzl:
```diff
@@ -88,7 +88,7 @@ def define_targets(rules):
         aten_ufunc_generated_cpu_sources() +
         aten_ufunc_generated_cpu_kernel_sources() + [
             "Declarations.yaml",
-        ] + gen_aten_outs_cuda + gen_aten_outs_mtia
```
Instead of directly adding mtia files to `gen_aten_outs`, I'm separating it out and use by dedicated genrule `gen_aten_mtia`. This is needed because this PR is introducing the `--mtia` flag to trochgen.

andyanwang committed:
[MTIA Aten Backend][2/n] Migrate clamp ops(clamp.out/clamp_min.out/clamp_max.out) from out-of-tree to in-tree (#154015)

Summary:

# Context

See the first PR https://github.com/pytorch/pytorch/pull/153670

# This diff 
Migrate clamp ops from out-of-tree to in-tree, including
- clamp.out
- clamp_min.out
- clamp_max.out

I had to migrate these 3 ops together, because clamp.out calls all 3 stubs(clamp_scalar_stub/clamp_min_scalar_stub/clamp_max_scalar_stub), which are also called by the other 2 ops.

Differential Revision: D74674418

andyanwang commented on build.bzl:
```diff
@@ -72,7 +72,7 @@ def define_targets(rules):
         "--source-path aten/src/ATen",
         "--aoti_install_dir=$(RULEDIR)/torch/csrc/inductor/aoti_torch/generated"
-    ] + (["--static_dispatch_backend CPU"] if rules.is_cpu_static_dispatch_build() else []))
+    ] + (["--static_dispatch_backend CPU"] if rules.is_cpu_static_dispatch_build() else []) + ["--mtia"])
```
n00b question, why this `build.bzl` file is exported to OSS? Isn't it a buck thing that's meta internal?

albanD replied:
It is so we can make such change as a pure github PR without have to codev with internal.
It might also be used for oss bazel build? Not sure.

andyanwang replied:
Thanks for the explanation

albanD reviewed (approved):
SGTM

nautsimon reviewed (approved):
awesome!
